components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentsUpdaterTest.kt

```diff
@@ -0,0 +1,90 @@
+import org.junit.Before
+import org.junit.Test
+import org.junit.Assert.assertFalse
+import org.junit.Assert.assertTrue
+import org.junit.runner.RunWith
+import org.mockito.Mockito.`when`
+import org.mockito.Mockito.spy
+import org.robolectric.RobolectricTestRunner
+
+@RunWith(RobolectricTestRunner::class)
+class ExperimentsUpdaterTest {
+    private var context: Context = ApplicationProvider.getApplicationContext()
+    private var glean: GleanInternalAPI = mock()
+    private lateinit var configuration: Configuration
+    private lateinit var experiments: ExperimentsInternalAPI
+    private lateinit var experimentStorage: FlatFileExperimentStorage
+    private lateinit var experimentSource: KintoExperimentSource
+    private lateinit var experimentsUpdater: ExperimentsUpdater
+
+    private fun getDefaultExperimentStorageMock(): FlatFileExperimentStorage {
+        val storage: FlatFileExperimentStorage = mock()
+        `when`(storage.retrieve()).thenReturn(ExperimentsSnapshot(emptyList(), null))
+        return storage
+    }
+
+    private fun getDefaultExperimentSourceMock(): KintoExperimentSource {
+        val source: KintoExperimentSource = mock()
+        `when`(source.getExperiments(mozilla.components.support.test.any())).thenReturn(ExperimentsSnapshot(emptyList(), null))
+        return source
+    }
+
+    private fun resetExperiments(
+        source: KintoExperimentSource = getDefaultExperimentSourceMock(),
+        storage: FlatFileExperimentStorage = getDefaultExperimentStorageMock(),
+        valuesProvider: ValuesProvider = ValuesProvider()
+    ) {
+        // Initialize WorkManager (early) for instrumentation tests.
+        WorkManagerTestInitHelper.initializeTestWorkManager(context)
+
+        configuration = Configuration()
+        experiments = spy(ExperimentsInternalAPI())
+        experiments.valuesProvider = valuesProvider
+
+        `when`(glean.isInitialized()).thenReturn(true)
+        `when`(experiments.getGlean()).thenReturn(glean)
+        experimentStorage = storage
+        `when`(experiments.getExperimentsStorage(context)).thenReturn(storage)
+
+        experimentsUpdater = spy(ExperimentsUpdater(context, experiments))
+        `when`(experiments.getExperimentsUpdater(context)).thenReturn(experimentsUpdater)
+        experimentSource = source
+        `when`(experimentsUpdater.getExperimentSource(configuration)).thenReturn(experimentSource)
+    }
+
+    @Before
+    fun setUp() {
+        resetExperiments()
+    }
+
+    @Test
+    fun `ExperimentsUpdater initialize schedules ExperimentsUpdateWorker in WorkManager`() {
+        assertFalse("ExperimentsUpdateWorker should not be scheduled yet",
+            isWorkScheduled(ExperimentsUpdater.TAG))
+        experimentsUpdater.initialize(configuration)
+        assertTrue("initialize must schedule ExperimentsUpdateWorker",
+            isWorkScheduled(ExperimentsUpdater.TAG))
+    }
+
+    @Test
+    fun `ExperimentsUpdateWorker doWork returns SUCCESS`() {
```

Comment:
What about test coverage for the failure case?

Reply:
It's actually hard to test that bit because of the call chain.  Let me take a stab at mocking the updater function to return a false and see if I can get coverage for that line.